decimal/decimal.go

```diff
@@ -0,0 +1,89 @@
+package decimal
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+)
+
+// Decimal represents a decimal amount, internally stored as a float64, with a
+// precision which determines how many decimal points are used when the float64
+// is converted to a string.
+type Decimal struct {
+	f         float64
+	precision int
+}
+
+// New returns a new Decimal.
+func New(f float64, precision int) Decimal {
+	return Decimal{f: f, precision: precision}
+}
+
+// MarshalJSON converts the Decimal to JSON bytes.
+func (d Decimal) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + d.String() + `"`), nil
+}
+
+// UnmarshalJSON reads JSON bytes into a Decimal. The precision is calculated
+// from the number of digits after the decimal point.
+func (d *Decimal) UnmarshalJSON(b []byte) error {
+	s := strings.Trim(string(b), `"`)
+	if strings.IndexByte(s, 'e') > -1 {
+		return errors.New("luno: unsupported decimal notation")
+	}
+	f, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		return err
+	}
+	*d = New(f, getPrecision(s))
+	return nil
+}
+
+func (d Decimal) String() string {
+	return strconv.FormatFloat(d.f, 'f', d.precision, 64)
+}
+
+func (d Decimal) Float64() float64 {
+	return d.f
+}
+
+// Sign returns the sign of the Decimal (-1 for negative, 1 for positive and
+// 0 for zero).
+func (d Decimal) Sign() int {
+	if d.f < 0 {
+		return -1
+	}
+	if d.f > 0 {
+		return 1
+	}
+	return 0
+}
+
+// Neg returns the negative of this Decimal.
+func (d Decimal) Neg() Decimal {
+	return Decimal{
+		f:         -d.f,
+		precision: d.precision,
+	}
+}
+
+// Add adds the given Decimal to this Decimal and returns the result. The result
+// has the greater of the two Decimals' precisions.
+func (d Decimal) Add(d2 Decimal) Decimal {
+	precision := d.precision
+	if d2.precision > precision {
+		precision = d2.precision
+	}
+	return Decimal{
+		f:         d.f + d2.f,
```

Comment:
I think we need to round the result to precision here because we're using floats.

Reply:
Hmm, we do that anyway when we convert to string/marshal.

Reply:
I'm worried about the rounding errors accumulating over multiple Add operations. The streaming code checks if the decimal is exactly zero (Sign() == 0) so any accumulated error will cause that check to fail.

Maybe we should just switch to big.Int instead of float64. Then we don't need to worry about such issues.

Reply:
Yeah it's the reason I didn't implement .Cmp().